src/lib.rs

```diff
@@ -40,19 +42,358 @@ mod repl_helper;
+                    eprintln!("Error: {:?}", e_str);
+                    1
+                }
+            }
+        }
+
+        #[no_mangle]
+        pub unsafe extern "C" fn scryer_consult_module_string(machine: &mut Machine, input: *const c_char) -> i32 {
+            let result = std::panic::catch_unwind(|| {
+                let c_str: &CStr;
+                unsafe {
+                    c_str = CStr::from_ptr(input);
+                }
+                c_str.to_str().expect("Not a valid UTF-8 string")
+            });
+
+            match result {
+                Ok(r_str) => {
+                    machine.consult_module_string("facts", r_str.to_owned());
+                    0
+                },
+                Err(e_str) => {
+                    eprintln!("Error: {:?}", e_str);
+                    1
+                }
+            }
+        }
+
+        /// `scryer_run_query` runs a prolog query using a Prolog machine instance.
+        ///
+        /// This function accepts a mutable reference (`machine`) to an instance of the Prolog `Machine`
+        /// and an immutable C pointer (`input`) to a char which is to contain a valid Prolog query.
+        /// It attempts to run the query on the machine and return a serialized JSON string as a raw
+        /// C string pointer, with a "status" field and a "result" field.
+        /// If anything goes wrong, the "status" will be set to "error" and an "error" field is included.
+        /// However, if the operation is successful, the "status" is set to "ok" and the "result" field will contain the result of the query.
+        ///
+        /// # Safety
+        ///
+        /// This function contains unsafe blocks due to the use of raw pointers, but it is required
+        /// to operate across FFI boundaries. It is the caller's responsibility to ensure the validity
+        /// of `input`. The `input` should be a valid Null-Terminated `c_char` string.
+        ///
+        /// The returned C string is allocated on the heap and should be deallocated using the
+        /// `scryer_free_c_string()` function to prevent memory leaks.
+        ///
+        /// # Parameters
+        ///
+        /// - `machine`: a mutable reference to a `Machine` instance.
+        /// - `input`: an input in form of raw C `*const c_char` pointer.
+        ///
+        /// # Returns
+        ///
+        /// A raw pointer to a C string containing the serialized JSON result.
+        ///
+        /// # Expected Response Format
+        /// ```json
+        /// {
+        ///   "status": "ok",  // Can also be "error" or "panic"
+        ///   "result": [ // Only present if status is "ok"
+        ///     { ... }, // Each is a Map representing a query result.
+        ///   ],
+        ///   "error": "..." // Only present if status is "error"
+        /// }
+        #[no_mangle]
+        pub unsafe extern "C" fn scryer_run_query(machine: &mut Machine, input: *const c_char) -> *mut c_char {
```

Comment:
A mixture of my comment to `scryer_start_new_query_generator` and `scryer_run_query_generator` apply, I don't feel like repeating myself. 
But I find the JSON format description better here. Though I would expect the result to also be able to be true or false here?